File: apps/backend/app/contract/validation.py
"""docs/json_contract.md §12. Reject, do not repair."""

from __future__ import annotations

from app.contract.models import Contract, is_valid_key


def validate_contract(contract: Contract) -> list[dict]:
    out: list[dict] = []

    def bad(rule: str, where: str, message: str) -> None:
        out.append({"rule": rule, "where": where, "message": message})

    seen_ids: set[str] = set()
    seen_z: set[int] = set()
    seen_names: set[str] = set()

    for key, node in contract.components.items():
        if not is_valid_key(key):
            bad("V2", key, "key does not match {type}_{slug(name)}")
        if key != node.key:
            bad("V2", key, f"key disagrees with the node's own key {node.key!r}")
        if node.id in seen_ids:
            bad("V3", key, f"duplicate id {node.id!r}")
        seen_ids.add(node.id)
        if not 0.0 <= node.opacity <= 1.0:
            bad("V4", key, f"opacity {node.opacity} out of 0.0..1.0")
        if node.rect.w <= 0 or node.rect.h <= 0:
            bad("V5", key, f"rect must have positive size, got {node.rect.w}x{node.rect.h}")
        if isinstance(node.radius, (int, float)) and node.radius < 0:
            bad("V6", key, f"radius {node.radius} is negative")
        if node.type == "text" and node.text is None:
            bad("V7", key, "a text node must carry text")
        if node.type == "ellipse" and node.radius != "50%":
            bad("V10", key, 'an ellipse must have radius "50%"')
        if node.z in seen_z:
            bad("V12", key, f"duplicate z value {node.z}")
        seen_z.add(node.z)
        if not node.name.strip():
            bad("V13", key, "name is blank")
        folded = node.name.strip().lower()
        if folded in seen_names:
            bad("V14", key, f"duplicate name {node.name!r} in this scope")
        seen_names.add(folded)

    return out

File: packages/design-contract/src/main/kotlin/com/ikk/core/contract/Validation.kt
package com.ikk.core.contract

/** docs/json_contract.md §12. Reject, do not repair. */
sealed interface ValidationResult {
    data object Ok : ValidationResult
    data class Invalid(val violations: List<Violation>) : ValidationResult
}

data class Violation(val rule: String, val where: String, val message: String) {
    override fun toString() = "$rule at $where: $message"
}

object ContractValidator {

    private val KEY_PATTERN = Regex("^(rect|ellipse|text|image)_[A-Za-z0-9]+$")

    fun validate(contract: Contract): ValidationResult {
        val v = mutableListOf<Violation>()

        if (contract.schemaVersion != Contract.SCHEMA_VERSION) {
            v += Violation("V1", "root", "schemaVersion ${contract.schemaVersion} != ${Contract.SCHEMA_VERSION}")
        }
        if (contract.layout != "relative") {
            v += Violation("V1", "root", "layout \"${contract.layout}\" is not supported in v1")
        }

        val seenIds = mutableSetOf<String>()
        val seenZ = mutableSetOf<Int>()
        val seenNames = mutableSetOf<String>()

        for ((key, node) in contract.components) {
            if (!KEY_PATTERN.matches(key)) {
                v += Violation("V2", key, "key does not match {type}_{slug(name)}")
            }
            if (key != node.key) {
                v += Violation("V2", key, "key disagrees with the node's own key \"${node.key}\"")
            }
            if (!seenIds.add(node.id)) {
                v += Violation("V3", key, "duplicate id \"${node.id}\"")
            }
            if (node.opacity !in 0.0..1.0) {
                v += Violation("V4", key, "opacity ${node.opacity} out of 0.0..1.0")
            }
            if (node.rect.w <= 0 || node.rect.h <= 0) {
                v += Violation("V5", key, "rect must have positive size, got ${node.rect.w}x${node.rect.h}")
            }
            val r = node.radius
            if (r is Radius.Dp && r.dp < 0) {
                v += Violation("V6", key, "radius ${r.dp} is negative")
            }
            if (node is TextNode && node.text == null) {
                v += Violation("V7", key, "a text node must carry text")
            }
            if (node is ImageNode && node.source != null && node.source.ref.isBlank()) {
                v += Violation("V8", key, "an image source must have a non-blank ref")
            }
            if (node is EllipseNode && node.radius !is Radius.Full) {
                v += Violation("V10", key, "an ellipse must have radius \"50%\"")
            }
            if (node.name.isBlank()) {
                v += Violation("V13", key, "name is blank")
            }
            if (!seenNames.add(node.name.trim().lowercase())) {
                v += Violation("V14", key, "duplicate name \"${node.name}\" in this scope")
            }
            if (!seenZ.add(node.z)) {
                v += Violation("V12", key, "duplicate z value ${node.z}")
            }
        }

        return if (v.isEmpty()) ValidationResult.Ok else ValidationResult.Invalid(v)
    }
}
